lntest/itest/lnd_forward_interceptor_test.go

```diff
@@ -29,166 +31,136 @@ type interceptorTestCase struct {
+	chanPoint := ctx.openChannel(ctx.alice, ctx.bob, chanAmt)
+	ctx.openChannel(ctx.bob, ctx.carol, chanAmt)
+	defer ctx.closeChannels()
+
+	// Connect the interceptor.
+	interceptor, cancelInterceptor := ht.HtlcInterceptor(ctx.bob)
+
+	// Prepare the test cases.
+	req := &lnrpc.Invoice{ValueMsat: 1000}
+	addResponse := ht.AddInvoice(req, ctx.carol)
+	invoice := ht.LookupInvoice(ctx.carol, addResponse.RHash)
+	tc := &interceptorTestCase{
+		amountMsat: 1000,
+		invoice:    invoice,
+		payAddr:    invoice.PaymentAddr,
+	}
+
+	// We initiate a payment from Alice.
+	done := make(chan struct{})
+	go func() {
+		ctx.sendPaymentAndAssertAction(tc)
+
+		// Signal that all the payments have been sent.
+		close(done)
+	}()
+
+	// We start the htlc interceptor with a simple implementation that
+	// saves all intercepted packets. These packets are held to simulate a
+	// pending payment.
+	packet := ht.ReceiveHtlcInterceptor(interceptor)
+
+	// Here we should wait for the channel to contain a pending htlc, and
+	// also be shown as being active.
+	err := wait.NoError(func() error {
+		channel := ht.QueryChannelByChanPoint(ctx.bob, chanPoint)
+
+		if len(channel.PendingHtlcs) == 0 {
+			return fmt.Errorf("expect alice <> bob channel to " +
+				"have pending htlcs")
+		}
+		if channel.Active {
+			return nil
+		}
+
+		return fmt.Errorf("channel not active")
+	}, defaultTimeout)
+	require.NoError(
+		ctx.ht, err, "alice <> bob channel pending htlc never arrived",
+	)
+
+	// At this point we want to make bob's link send all pending htlcs to
+	// the switch again. We force this behavior by disconnecting and
+	// connecting to the peer.
+	ctx.ht.DisconnectNodes(ctx.bob, ctx.alice)
+	ctx.ht.EnsureConnected(ctx.bob, ctx.alice)
+
+	// Here we wait for the channel to be active again.
+	ht.AssertChannelExists(ctx.bob, chanPoint)
+
+	// Now that the channel is active we make sure the test passes as
+	// expected.
+
+	// We expect one in flight payment since we held the htlcs.
+	var preimage lntypes.Preimage
+	copy(preimage[:], invoice.RPreimage)
+	ht.AssertPaymentStatus(ctx.alice, preimage, lnrpc.Payment_IN_FLIGHT)
+
+	// At this point if we have more than one held htlcs then we should
+	// fail. This means we hold the same htlc twice which is a risk we want
+	// to eliminate. If we don't have the same htlc twice in theory we can
+	// cancel one and settle the other by mistake.
+	errDone := make(chan struct{})
+	go func() {
+		_, err := interceptor.Recv()
+		require.Error(ht, err, "expected an error from interceptor")
+
+		status, ok := status.FromError(err)
+		switch {
+		// If it is just the error result of the context cancellation
+		// the we exit silently.
+		case ok && status.Code() == codes.Canceled:
+			fallthrough
+
+		// When the test ends, during the node's shutdown it will close
+		// the connection.
+		case strings.Contains(err.Error(), "closed network connection"):
+			fallthrough
+
+		case strings.Contains(err.Error(), "EOF"):
+			close(errDone)
```

Comment:
Same here concerning `defer` and use of `require` inside goroutines.

Reply:
Fixed!